### PR TITLE
cifar10 dataset class

### PR DESCRIPTION
**Pull request by AvindaShamal**
Implemented a pytorch custom dataset class for cifar10 dataset and the first code included how to use preloaded dataset using torch.

### CODE REVIEW

File: avinda_shamal/src/task_1.1.py
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.1.py:1]: You can move this to the example folder

File: avinda_shamal/src/task_1.2.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    plt.imshow(img_)
+plt.show()
+
+## Another method to display image and label
+train_features, train_labels = next(iter(train_dataloader))
+img = train_features[0].squeeze()
+label = train_labels[0]
+plt.imshow(img.permute(1, 2, 0))
+plt.show()
+print(f"Label: {labels_map[label.item()]}")
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: This part can be moved into an example folder

File: avinda_shamal/src/task_1.2.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+import os
+import torch
+from torch.utils.data import Dataset
+from torchvision.io import read_image
+from torch.utils.data import DataLoader
+from torchvision import transforms
+import matplotlib.pyplot as plt
+
+class CustomImageDataset(Dataset):
+
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: remove this space

File: avinda_shamal/src/task_1.2.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+from torch.utils.data import DataLoader
+from torchvision import transforms
+import matplotlib.pyplot as plt
+
+class CustomImageDataset(Dataset):
+    
+    '''A custom dataset class for loading images from a directory structure where each class has its own subdirectory.'''
+    
+    def __init__(self, root_dir, transform=None, target_transform=None):
+
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: remove this space

File: avinda_shamal/src/task_1.2.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+from torch.utils.data import Dataset
+from torchvision.io import read_image
+from torch.utils.data import DataLoader
+from torchvision import transforms
+import matplotlib.pyplot as plt
+
+class CustomImageDataset(Dataset):
+    
+    '''A custom dataset class for loading images from a directory structure where each class has its own subdirectory.'''
+
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: remove this space

File: avinda_shamal/src/task_1.2.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        Initializes the CustomImageDataset.
+
+        Args:
+            root_dir (str): Path to the root directory containing class subdirectories.
+            transform (callable, optional): A function/transform to apply to the images (default: None).
+            target_transform (callable, optional): A function/transform to apply to the labels (default: None).
+            
+        Returns: None
+        """
+
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: remove this space

File: avinda_shamal/src/task_1.2.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+from torchvision.io import read_image
+from torch.utils.data import DataLoader
+from torchvision import transforms
+import matplotlib.pyplot as plt
+
+class CustomImageDataset(Dataset):
+    
+    '''A custom dataset class for loading images from a directory structure where each class has its own subdirectory.'''
+    
+    def __init__(self, root_dir, transform=None, target_transform=None):
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: please add type hints

File: avinda_shamal/src/task_1.2.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        """
+        Returns the total number of samples in the dataset.
+
+        Returns:
+            int: Number of samples in the dataset.
+        """
+        
+        return len(self.data)
+
+    def __getitem__(self, idx):
```
Comment (OshanJayawardena) [inline avinda_shamal/src/task_1.2.py]: please add type hints

**Review by OshanJayawardena — CHANGES_REQUESTED** [review]
Please use the ruff formatter addon on vscode to format your code. This will help you with proper spacing

 Add input and output type hints

restructure the code as `examples`, `src`, `tests`

File: avinda_shamal/src/cifar_class.py
Comment (OshanJayawardena) [inline avinda_shamal/src/cifar_class.py:1]: Please rename this as src/dataclasses.py

File: avinda_shamal/test/unit/test_cifar_len.py
Comment (OshanJayawardena) [inline avinda_shamal/test/unit/test_cifar_len.py:1]: Please remove testing scripts from this PR. This should be a separate PR

**Review by OshanJayawardena — CHANGES_REQUESTED** [review]
Two more changes and we can merge

File: avinda_shamal/src/cifar_class.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        
+        img_path, label = self.data[idx]
+        image = read_image(img_path)
+        if self.transform:
+            image = self.transform(image)
+        if self.target_transform:
+            label = self.target_transform(label)
+        return image, label
+
```
Comment (OshanJayawardena) [inline avinda_shamal/src/cifar_class.py]: Extra spaces are still there. Please use ruff formatter to format the files

**Review by OshanJayawardena — CHANGES_REQUESTED** [review]
- Unnecessary spaces are still there. 
- Please keep PRs short and precise. Do not convolute two tasks into a single PR.
- The testing should be a separate PR. 
- Once you have addressed a comment, please reply to that comment mentioning you have completed that. If you do not agree with the comment, you can reply your view.

File: avinda_shamal/src/task_1.1.py
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.1.py:1]: done it

File: avinda_shamal/src/task_1.2.py (hunk 1 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done this

File: avinda_shamal/src/task_1.2.py (hunk 2 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done

File: avinda_shamal/src/task_1.2.py (hunk 3 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done

File: avinda_shamal/src/task_1.2.py (hunk 6 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done

File: avinda_shamal/src/task_1.2.py (hunk 5 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done

File: avinda_shamal/src/task_1.2.py (hunk 4 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done

File: avinda_shamal/src/task_1.2.py (hunk 7 above)
Comment (AvindaShamal) [reply avinda_shamal/src/task_1.2.py]: done

File: avinda_shamal/test/unit/test_cifar_len.py
Comment (AvindaShamal) [reply avinda_shamal/test/unit/test_cifar_len.py:1]: done

File: avinda_shamal/src/cifar_class.py (hunk 8 above)
Comment (AvindaShamal) [reply avinda_shamal/src/cifar_class.py]: done

File: avinda_shamal/src/cifar_class.py
Comment (AvindaShamal) [reply avinda_shamal/src/cifar_class.py:1]: done it

**Comment by AvindaShamal** [comment]
Done everything

File: avinda_shamal/examples/hf_model_example.py
Comment (OshanJayawardena) [inline avinda_shamal/examples/hf_model_example.py:1]: This should be a seperate PR

File: avinda_shamal/src/package_avinda_shamal.egg-info/PKG-INFO
Comment (OshanJayawardena) [inline avinda_shamal/src/package_avinda_shamal.egg-info/PKG-INFO:1]: what is this?

File: avinda_shamal/examples/hf_model_example.py
Comment (AvindaShamal) [reply avinda_shamal/examples/hf_model_example.py:1]: added them to a new branch

File: avinda_shamal/src/package_avinda_shamal.egg-info/PKG-INFO
Comment (AvindaShamal) [reply avinda_shamal/src/package_avinda_shamal.egg-info/PKG-INFO:1]: removed them

**Review by OshanJayawardena — APPROVED** [review]
Good to go, you can merge now
